sdk/core/Azure.Core/src/Shared/OperationInternal.cs

```diff
@@ -0,0 +1,236 @@
+
+        protected ClientDiagnostics ClientDiagnostics { get; }
+
+        protected RequestFailedException OperationFailedException { get; set; }
+
+        public async ValueTask<Response> UpdateStatusAsync(CancellationToken cancellationToken) =>
+            await UpdateStatusAsync(async: true, cancellationToken).ConfigureAwait(false);
+
+        public Response UpdateStatus(CancellationToken cancellationToken) =>
+            UpdateStatusAsync(async: false, cancellationToken).EnsureCompleted();
+
+        public async ValueTask<Response<TResult>> WaitForCompletionAsync(CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                Response response = await UpdateStatusAsync(cancellationToken).ConfigureAwait(false);
+
+                if (HasCompleted)
+                {
+                    return Response.FromValue(Value, response);
+                }
+
+                var pollingInterval = response.Headers.TryGetValue(RetryAfterHeaderName, out string retryAfterValue)
+                    && int.TryParse(retryAfterValue, out int delayInSeconds)
+                    ? TimeSpan.FromSeconds(delayInSeconds) : DefaultPollingInterval;
+
+                await Task.Delay(pollingInterval, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        public async ValueTask<Response<TResult>> WaitForCompletionAsync(TimeSpan pollingInterval, CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                Response response = await UpdateStatusAsync(cancellationToken).ConfigureAwait(false);
+
+                if (HasCompleted)
+                {
+                    return Response.FromValue(Value, response);
+                }
+
+                await Task.Delay(pollingInterval, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        protected virtual async ValueTask<Response> UpdateStateAsync(bool async, CancellationToken cancellationToken)
+        {
+            Response response = async
+                ? await _operation.GetResponseAsync(cancellationToken).ConfigureAwait(false)
+                : _operation.GetResponse(cancellationToken);
+
+            var state = _operation.UpdateState(response);
+
+            SetState(state, response);
+
+            return response;
+        }
+
+        protected void SetState(OperationState<TResult> state, Response response)
+        {
+            lock (SetStateLock)
+            {
+                RawResponse = response;
+
+                if (state.Succeeded == true)
+                {
+                    Value = state.Value;
+                    HasCompleted = true;
+                }
+                else if (state.Succeeded == false)
+                {
+                    OperationFailedException = state.OperationFailedException;
+                    HasCompleted = true;
+
+                    throw OperationFailedException;
```

Comment:
Do we throw from UpdateStatus when an operation fails?

Reply:
Yes. It's a behavior we have in multiple LROs to let the user know the operation has failed. We also store the exception so we can throw it again in case someone calls the property `Value` later.